vim*: organize

khaneliman opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done
Follow up to https://github.com/NixOS/nixpkgs/pull/390702
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

khaneliman committed:
vimPlugins: move extras to separate files per source

Make it more obvious where certain plugins come from.

khaneliman committed:
vimUtils: move hooks to new directory

Add some organization

khaneliman committed:
vimUtils: move to utils directory

Add some organization

khaneliman committed:
vimPlugins.vim2nix: move to non-generated

khaneliman committed:
vimPlugins.markdown-preview-nvim: move supporting files to patches

MattSturgeon commented on pkgs/applications/editors/vim/plugins/luaPackagePlugins.nix:
```diff
@@ -0,0 +1,48 @@
+      "telescope-manix"
+      "telescope-nvim"
+    ];
+    toVimPackage =
+      name:
+      buildNeovimPlugin {
+        luaAttr = luaPackages.${name};
+      };
```
At this point, we could probably just inline this into the `genAttrs` expression? I don't think having a `toVimPackage` binding adds any readability anymore.

```nix
lib.genAttrs luarocksPackageNames (
  name:
  buildNeovimPlugin {
    luaAttr = luaPackages.${name};
  }
)
```

MattSturgeon commented on pkgs/applications/editors/vim/plugins/nodePackagePlugins.nix:
```diff
@@ -63,42 +58,3 @@ self: super:
   in
   lib.genAttrs nodePackageNames nodePackage2VimPackage
```
Same, may as well inline `nodePackage2VimPackage` given this file is now dedicated to node packages, I don't think having a binding adds any readability anymore.

MattSturgeon commented on pkgs/applications/editors/vim/plugins/overrides.nix:
```diff
@@ -3519,7 +3519,7 @@ in
   vim2nix = buildVimPlugin {
     pname = "vim2nix";
     version = "1.0";
-    src = ./vim2nix;
+    src = ./non-generated/vim2nix;
     dependencies = [ self.vim-addon-manager ];
   };
```
Should package definitions that aren't "overriding" anything be move out of `overrides.nix` and into another file?

I wonder if we could take a semi-automated approach, similar in principal to other by-name trees, where each directory in `non-generated` gets imported into `vimPlugins` via `callPackage`, using the directory name as the plugin attr name.

khaneliman replied:
Yeah, I think we should basically have a non-generated.nix or something that houses them instead of being inside overrides.

khaneliman replied:
Started implementing it. Didn't realize how many we had in there that were new plugin definitions in overrides.nix

MattSturgeon reviewed (commented):
I like the overall direction. What's needed to undraft this?

khaneliman commented:
> I like the overall direction. What's needed to undraft this?

I just wanted to double check on what we wanted to do with handling those overrides that might be unnecessary. Was going to either:

1. Drop them completely and we try to fix nixvim asap
2. Move them to the `luaPackagePlugins` file, until we know if they are safe to drop.

khaneliman committed:
vimPlugins: automate non-generated plugin calls

Move the non-generated plugins to using a function call that iterates
over the folders in the non-generated directory so we don't need them in
overrides.nix

khaneliman committed:
vimPlugins.Spacegray-vim: drop package

khaneliman committed:
vimPlugins.vim-pony: redirect to active repo

khaneliman committed:
vimplugins.coc-nginx: move overrides into nodePackagePlugins

khaneliman committed:
vimplugins.coc-*: move into cocPlugins

coc is breaking out packages into top-level, but I don't know if we
really need separate folders for each of these. Can just create another
set to merge in.

MattSturgeon commented:
> I just wanted to double check on what we wanted to do with handling those overrides that might be unnecessary. Was going to either:
> 
>     1. Drop them completely and we try to fix nixvim asap
> 
>     2. Move them to the `luaPackagePlugins` file, until we know if they are safe to drop.

I'm hoping https://github.com/nix-community/nixvim/pull/3099 will fix the nixvim issue. I've tested it against e8d0b02af0958823c955aaab3c82b03f54411d91 and it _seems_ to be working.

khaneliman commented:
> > I just wanted to double check on what we wanted to do with handling those overrides that might be unnecessary. Was going to either:
> > ```
> > 1. Drop them completely and we try to fix nixvim asap
> > 
> > 2. Move them to the `luaPackagePlugins` file, until we know if they are safe to drop.
> > ```
> 
> I'm hoping [nix-community/nixvim#3099](https://github.com/nix-community/nixvim/pull/3099) will fix the nixvim issue. I've tested it against [e8d0b02](https://github.com/NixOS/nixpkgs/commit/e8d0b02af0958823c955aaab3c82b03f54411d91) and it _seems_ to be working.

Want me to drop it from here and then test against this branch?

MattSturgeon commented:
> Want me to drop it from here and then test against this branch?

I'm conscious that this PR is growing quickly in scope, but sure.

I guess specific parts of the refactoring can be split into separate PRs if needed.

khaneliman committed:
vimPlugins: drop luaPackage overrides

Attempt to remove overrides that provide dependencies that should be
handled from the luaPackages derivation and buildNeovimPlugin.

khaneliman committed:
vimPlugins.plenary-nvim: migrate to luaPackagePlugins

khaneliman commented:
> > Want me to drop it from here and then test against this branch?
> 
> I'm conscious that this PR is growing quickly in scope, but sure.
> 
> I guess specific parts of the refactoring can be split into separate PRs if needed.

Yeah, I can definitely split stuff out if we need to.

MattSturgeon commented on pkgs/applications/editors/vim/plugins/nodePackagePlugins.nix:
```diff
@@ -0,0 +1,60 @@
+  nodePackages,
+}:
+self: super:
+(
```
Same

PerchunPak reviewed (approved):
Is there a reason this is still draft?

khaneliman commented:
> Is there a reason this is still draft?

Mostly just because I whipped out real quick without too much thought and wanted to give myself a chance to form a better view of what the goal was and if it was being done right. 

Also, needed to test the luaPackage override changes a bit so I could feel more comfortable with them.

khaneliman commented on pkgs/applications/editors/vim/plugins/default.nix:
```diff
@@ -27,6 +27,12 @@ let
 
+  nonGeneratedPlugins =
+    self: super:
+    lib.mapAttrs (name: _: callPackage (./non-generated + "/${name}") { }) (
+      lib.filterAttrs (name: type: type == "directory") (builtins.readDir ./non-generated)
+    );
```
Yeah, I think that's a literal more natural to read.

khaneliman commented on pkgs/applications/editors/vim/plugins/nodePackagePlugins.nix:
```diff
@@ -47,6 +47,7 @@ self: super:
       "coc-wxml"
       "coc-yaml"
       "coc-yank"
+      "@yaegassy/coc-nginx"
```
Ah, good spot.

khaneliman commented on pkgs/applications/editors/vim/plugins/overrides.nix:
```diff
@@ -1383,24 +1378,6 @@ in
-      nvim-treesitter-parsers.norg
-    ];
-
-    # Add magick to package.path
-    patches = [ ./patches/image-nvim/magick.patch ];
-
-    postPatch = ''
-      substituteInPlace lua/image/magick.lua \
-        --replace-fail @nix_magick@ ${luajitPackages.magick}
-    '';
```
This is where we are expecting the `luaPackages` set to work properly and need to verify it does. 

```nix
  image-nvim = prev.image-nvim.overrideAttrs (oa: {
    propagatedBuildInputs = [
      lua
      luajitPackages.magick
    ];
  });
```

khaneliman commented on pkgs/applications/editors/vim/plugins/overrides.nix:
```diff
@@ -2794,18 +2758,6 @@ in
-  rest-nvim = super.rest-nvim.overrideAttrs {
-    dependencies = with self; [
-      plenary-nvim
-      (nvim-treesitter.withPlugins (p: [
-        p.http
-        p.json
-      ]))
```
So, not sure if rockspec is accurate... they don't list `plenary-nvim` as a dep and it's not in overrides in lua-modules. But, we had to add that to get the neovim-require-check-hook to pass when it was built with `buildVimPlugin`

teto commented on pkgs/applications/editors/vim/plugins/non-generated/Spacegray-vim/default.nix:
```diff
@@ -0,0 +1,20 @@
+  pname = "Spacegray.vim";
+  version = "2021-07-06";
+  src = fetchFromGitHub {
+    owner = "ackyshake";
```
https://github.com/alek3y/spacegray.vim seems archived

khaneliman commented on pkgs/applications/editors/vim/plugins/non-generated/VimCompletesMe/default.nix:
```diff
@@ -0,0 +1,19 @@
+  pname = "VimCompletesMe";
+  version = "2022-02-18";
+  src = fetchFromGitHub {
+    owner = "ackyshake";
```
Good find :) Gonna try adding it back to list with that url

teto commented on pkgs/applications/editors/vim/plugins/non-generated/bitbake-vim/default.nix:
```diff
@@ -0,0 +1,21 @@
+    sha256 = "1hfly2vxhhvjdiwgfz58hr3523kf9z71i78vk168n3kdqp5vkwrp";
+  };
+
+  meta.homepage = "https://github.com/sblumentritt/bitbake.vim/";
```
doesnt seem reachable. It was nice to provide this thanks to nix cache but if we can't build from source and you are doing a cleanup anyway drop it. If you feel nice you can suggest an alternative (i dunno any).

khaneliman replied:
Should we redirect this to https://github.com/kergoth/vim-bitbake ?

teto replied:
I think bitbake users should use https://github.com/openembedded/bitbake/tree/master/contrib/vim instead. If you feel motivated, you could change the src by that url instead

teto commented on pkgs/applications/editors/vim/plugins/non-generated/vim-pony/default.nix:
```diff
@@ -0,0 +1,19 @@
+    owner = "jakwings";
+    repo = "vim-pony";
+    rev = "b26f01a869000b73b80dceabd725d91bfe175b75";
+    sha256 = "0if8g94m3xmpda80byfxs649w2is9ah1k8v3028nblan73zlc8x8";
```
unreachable .  drop it

khaneliman replied:
map to https://github.com/jmcomets/vim-pony/ ?

teto replied:
I dont know what vim-pony did but if they are the same why not.

khaneliman committed:
vimPlugins.VimCompletesMe: migrate to active repo

Old repo referenced was removed, referencing currently active repo.

teto commented on pkgs/applications/editors/vim/plugins/default.nix:
```diff
@@ -19,12 +19,27 @@ let
     inherit buildVimPlugin;
+  };
+
+  luaPackagePugins = callPackage ./luaPackagePlugins.nix {
```
Plugins not Pugins :)

khaneliman replied:
Doh.. lol

khaneliman commented on pkgs/applications/editors/vim/plugins/nodePackagePlugins.nix:
```diff
@@ -0,0 +1,59 @@
+  buildVimPlugin,
+  nodePackages,
+}:
+self: super:
```
That's fine with me I can update them

teto commented:
There seems to be unnecessary renaming like the `vim2nix` folder or the `markdown-preview.nvim` (the renaming is worse imo). the moving of the hooks or vim-utils.nix.

nixpkgs is used by many people, let's not rename for the sake of renaming. It can break PRs, forks etc. I know it will break mine, the nixpkgs updater.

I also dont like the negative in `non-generated` path. Some of them can be "generated" via nix-update or not. 

I kinda liked having all overrides in one file, I found it made it easier to contribute because you knew you just had to look in one place. I fear this makes contributing harder but I can understand why you would want coc packages and lua-based packages in their own files. So maybe to help contributors, reference this cahnge in the doc or at the top of the overrides.nix ?

MattSturgeon commented:
> There seems to be unnecessary renaming like the `vim2nix` folder or the `markdown-preview.nvim` (the renaming is worse imo). the moving of the hooks or vim-utils.nix.
> 
> nixpkgs is used by many people, let's not rename for the sake of renaming. It can break PRs, forks etc. I know it will break mine, the nixpkgs updater.

I can sympathies with this, but I don't think we should shy away from renaming things most contributors agree the new name/structure/layout is better.

Maybe this would be easier to judge and review if each of the reorganizations was moved to a dedicated PR?


> I also dont like the negative in `non-generated` path. Some of them can be "generated" via nix-update or not.

I agree we could come up with a better name, although I think the negativity is intended to encourage people to contribute plugins using the usual "generated" approach. The non-generated plugins _should_ be the exception, so a negative name makes some sense to me.

Not sure what you mean by some of them can be generated. Surely any that can be generated would be moved _out_ of the non-generated tree and into `generated.nix`?

> I kinda liked having all overrides in one file, I found it made it easier to contribute because you knew you just had to look in one place.

I think we still have all overrides in one file. The other files are not overriding anything, but rather adding entire packages.

IMO it was more confusing before, having the overrides file contain a mix of overrides and package definitions.

khaneliman commented:
> There seems to be unnecessary renaming like the `vim2nix` folder or the `markdown-preview.nvim` (the renaming is worse imo). the moving of the hooks or vim-utils.nix.
>
> nixpkgs is used by many people, let's not rename for the sake of renaming. It can break PRs, forks etc. I know it will break mine, the nixpkgs updater.

The just moved folders so they get placed in appropriate locations for what they are. It's not for nothing, I would think. 
 
> I also dont like the negative in `non-generated` path. Some of them can be "generated" via nix-update or not.
> 

Yeah, I agree I don't like the name either... would be open for suggestions from everyone.

> I kinda liked having all overrides in one file, I found it made it easier to contribute because you knew you just had to look in one place. I fear this makes contributing harder but I can understand why you would want coc packages and lua-based packages in their own files. So maybe to help contributors, reference this change in the doc or at the top of the overrides.nix ?

Yeah, that's fair.. but, they weren't really overrides in the same sense as the others in that file. They were just plugin definitions that weren't part of an update script generating them and are built from existing derivations. I can put a note in docs when we solidify naming / locations.

MattSturgeon commented on pkgs/applications/editors/vim/plugins/luaPackagePlugins.nix:
```diff
@@ -0,0 +1,48 @@
+  luaPackages = neovim-unwrapped.lua.pkgs;
+in
+self: super:
+(
```
Looks like this was reverted by mistake last force-push?

MattSturgeon commented on pkgs/applications/editors/vim/plugins/overrides.nix:
```diff
@@ -283,19 +283,6 @@ in
     ];
   };
 
-  # The GitHub repository returns 404, which breaks the update script
```
Was this reverted intentionally because a new repo has been found?

khaneliman replied:
Sorry, thought I added them to new definitions.. but might be dropping the plugins if we don't point to the new repos I found.

teto commented:
> Yeah, that's fair.. but, they weren't really overrides in the same sense as the others in that file. They were just plugin definitions that weren't part of an update script generating them and are built from existing derivations. I can put a note in docs when we solidify naming / locations.

Right in that case, it's a better to have them in separate files since they were not overrides. Thanks for the cleanup.

khaneliman committed:
vimPlugins.bitbake-vim: migrate to new repo and move

Move to actively maintained repo and move to non-generated since plugin
updater can't handle sub directories.

khaneliman commented:
I think I'm pleased with overall organization and most of the cleanup. Main things that need to be verified is testing the luaPackages that had the overrides removed.

GaetanLepage reviewed (commented):
Thank you @khaneliman! That's more than welcomed.

Also thanks to @MattSturgeon and @teto for the thorough review.
I have looked at each commit and I am fine with the changes made.
All good on my end :)

Tests:
- [x] `nixpkgs-review` on all platforms
- [x] `nixpkgs-review -p vimPlugins` on `x86_64-linux`
- [x] [Nixvim test suite](https://github.com/nix-community/nixvim/pull/3108)

teto commented:
It's a nice PR but I am still unsure about moving the files to "non-generated" or to "utils/". Is it a problem to leave them where they are ? especially if you dont like the name either. 
I find `plugins/vim2nix` more meaningful than `plugins/non-generated/vim2nix`, generated things in nixpkgs are usually called "generated" so they are not by default. 
Wont moving the updater conflict with https://github.com/NixOS/nixpkgs/pull/380691 ?

teto commented:
I apologize I had not noticed we already had a "non-generated" folder. I thought you were introducing one. Ok for those.
Let's go
